golang/cosmos/x/swingset/types/msgs.go

```diff
@@ -48,6 +51,52 @@ func chargeAdmission(ctx sdk.Context, keeper SwingSetKeeper, addr sdk.AccAddress
 	return keeper.ChargeBeans(ctx, addr, beans)
 }
 
+// checkSmartWalletProvisioned verifies if a smart wallet message can be
+// delivered for the owner's address.
+func checkSmartWalletProvisioned(ctx sdk.Context, keeper SwingSetKeeper, addr sdk.AccAddress) (sdk.Context, error) {
+	// While multiple swingset messages are not currently allowed in the same
+	// transaction, if that is ever relaxed, we would end up generating multiple
+	// provision actions, and the provisioning code is not fully idempotent.
+	_, txShouldProvision := ctx.Context().Value(TxShouldProvision).(MsgProvision)
+	if txShouldProvision {
+		return ctx, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "cannot auto-provision for multiple smart wallet messages")
+	}
+
+	// This checks if a smart wallet has already been provisioned by the controller
+	// However a provision (either explicit or automatic) may be pending execution
+	// and this check will not catch that case, resulting in the owner being charged
+	// for provisioning again.
+	// TODO: mark that a smart wallet provision is pending. However in that case,
+	// auto-provisioning should still be performed (but without fees being charged),
+	// until the controller actually provisions the smart wallet (the operation may
+	// transiently fail, requiring retries until success)
+	isProvisioned, err := keeper.HasSmartWallet(ctx, addr)
+	if err != nil {
+		return ctx, err
+	} else if isProvisioned {
+		return ctx, nil
+	}
+
+	beansPerUnit := keeper.GetBeansPerUnit(ctx)
+	beans := beansPerUnit[BeansPerSmartWalletProvision]
+
+	// This is a separate charge from the smart wallet action which triggered the check
+	err = keeper.ChargeBeans(ctx, addr, beans)
+	if err != nil {
+		return ctx, err
+	}
```

Comment:
I'd appreciate it if you changed `HasSmartWallet` to `GetSmartWalletState`, returning an enum that is one of `SmartWalletUnspecified`, `SmartWalletNone`, `SmartWalletPending`, and `SmartWalletProvisioned`.

Then encapsulate the charging behavior and simplify the comment above this code to indicate what is expected. 
 The header comment above `keeper.ChargeForSmartWallet` should explain what it wants to do versus what is actually implemented.

```suggestion
	walletState, err := keeper.GetSmartWalletState(ctx, addr)
	if err != nil {
		return ctx, err
	} else if walletState == SmartWalletProvisioned {
		// No need to do anything; they already have a smart wallet.
		return ctx, nil
	}

	// Only charge to move to the pending state.
	if walletState != SmartWalletPending {
		// TODO: this call should actually transition to SmartWalletPending on success.
		err = keeper.ChargeForSmartWallet(ctx, addr)
		if err != nil {
			return ctx, err
		}
	}
```

Reply:
> keeper.ChargeForSmartWallet

I was hoping to avoid yet another method on the `SwingSetKeeper` interface, but I suppose we'd need to have something to mark a pending smart wallet if/when we do that. It does feel a little weird to have some of the fee calculations in admission, some in the msg server, and now, some on the keeper as well.

Reply:
I don't really care where that method/function lives, just that it encapsulates the charging behaviour, so you could pass the keeper as an argument instead.  It can be a private helper function (starts with lowercase).